### PR TITLE
Granular catchups

### PR DESCRIPTION
**Pull request by jonsequitur**

### CODE REVIEW

File: Alluvial.Tests/(Infrastructure)/Wait.cs [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+namespace Alluvial.Tests
+{
+    public static class Wait
+    {
+        public static async Task Until(
+            Func<bool> until,
+            TimeSpan? pollInterval = null,
+            TimeSpan? timeout = null)
+        {
```
Comment (nathanboktae) [inline Alluvial.Tests/(Infrastructure)/Wait.cs]: This takes like 3 lines to do in Rx right?

File: Alluvial.Tests/Alluvial.Tests.csproj [hunk 2]
```diff
@@ -86,7 +101,7 @@
     <Compile Include="Properties\AssemblyInfo.cs" />
     <Compile Include="StreamQueryTests.cs" />
     <Compile Include="TracingTests.cs" />
-    <Compile Include="Wait.cs" />
+    <Compile Include="%28Infrastructure%29\Wait.cs" />
```
Comment (nathanboktae) [inline Alluvial.Tests/Alluvial.Tests.csproj]: Not a big fan of parenthesis as I can't just copy and paste them from a `git status`

File: Alluvial.Tests/(Infrastructure)/Wait.cs (hunk 1 above)
Comment (jonsequitur) [inline Alluvial.Tests/(Infrastructure)/Wait.cs]: If you want it fully async it's not entirely straightforward. For imperative tasks, async / await is a little more readable I think.

File: Alluvial.Tests/StreamCatchupTests.cs [hunk 3]
```diff
@@ -191,27 +134,24 @@ await StreamCatchup.Create(streamSource.UpdatedStreams())
         public async Task When_projections_are_cursors_then_catchup_does_not_replay_previously_seen_events()
         {
             var projectionStore = new InMemoryProjectionStore<BalanceProjection>();
```
Comment (nathanboktae) [inline Alluvial.Tests/StreamCatchupTests.cs]: I'm failing to see how this tests this scenario. Shouldn't you run the catch up twice? Or provide the projection at an initial cursor state?

Comment (jonsequitur) [inline Alluvial.Tests/StreamCatchupTests.cs]: Yeah, I think something got lost in translation there. I'll fix this.

File: Alluvial.Tests/StreamCatchupTests.cs [hunk 4]
```diff
@@ -278,82 +212,46 @@ await StreamCatchup.Create(streamSource.UpdatedStreams())
...
-                                               configure: c => c.StoreCursor(async (id, cursor) => storedCursor = cursor))
-                                       .Subscribe(new BalanceProjector());
-
-            var query = await catchup.RunSingleBatch();
-
-            storedCursor.Should().BeSameAs(query.Cursor);
-        }
-
-        [Test]
-        public async Task Catchup_cursor_retrieval_can_be_specified_using_catchup_configuration()
```
Comment (nathanboktae) [inline Alluvial.Tests/StreamCatchupTests.cs]: Is this not a scenario anymore? seems that it is.

File: Alluvial.Tests/StreamOfStreamsCatchupTests.cs [hunk 5]
```diff
@@ -0,0 +1,485 @@
...
+                                                      .Pipeline(async (projection, batch, next) =>
+                                                      {
+                                                          await Task.Delay(500);
+                                                          await next(projection, batch);
+                                                      }), projectionStore);
+
+            catchup.RunSingleBatch();
+            catchup.RunSingleBatch();
+
+            await Task.Delay(1000);
```
Comment (nathanboktae) [inline Alluvial.Tests/StreamOfStreamsCatchupTests.cs]: Probably a better idea to use `Task.WaitAll` to wait on the previous two tasks

File: Alluvial.Tests/StreamOfStreamsCatchupTests.cs [hunk 6]
```diff
@@ -0,0 +1,485 @@
...
+        {
+            var projectionStore = new InMemoryProjectionStore<BalanceProjection>();
+
+            var catchup = StreamCatchup.Create(streamSource.UpdatedStreams(), batchCount: 50)
+                                       .Subscribe(new BalanceProjector(), projectionStore);
+
+            using (catchup.Poll(TimeSpan.FromMilliseconds(10)))
+            {
+                // write more events
+                Task.Run(async () =>
```
Comment (nathanboktae) [inline Alluvial.Tests/StreamOfStreamsCatchupTests.cs]: use `Background.Poll` here too?

File: Alluvial.Tests/StreamCatchupTests.cs (hunk 4 above)
Comment (jonsequitur) [inline Alluvial.Tests/StreamCatchupTests.cs]: In the case of a single-stream catchup, the projection is the cursor, so it's not really necessary to pull this out as a separate concept. I'm going to leave this out of the interface for now and bring it back if we find it's needed.

**Pull request merged**